src/irclog2html/irclogsearch.py

```diff
@@ -236,39 +243,89 @@ def print_search_results(query, where=None):
+        print("</ul>", file=stream)
     total_time = time.time() - started
     print("<p>%d matches in %d log files with %d lines (%.1f seconds).</p>" % (
-                stats.matches, stats.files, stats.lines, total_time))
-    print(FOOTER)
+                stats.matches, stats.files, stats.lines, total_time),
+          file=stream)
+    print(FOOTER, file=stream)
+
+    return formatter # destroying it closes the stream, breaking the result
 
 
 def rewrap_stdout():
-    # XXX: shouldn't replace sys.stdout, should pass the output stream to
-    # the functions that want to print stuff
     if hasattr(sys.stdout, 'buffer'):
         stream = sys.stdout.buffer # Python 3
     else:
         stream = StdoutWrapper(sys.stdout) # Python 2
-    sys.stdout = io.TextIOWrapper(stream, 'ascii',
-                                  errors='xmlcharrefreplace',
-                                  line_buffering=True)
+    return io.TextIOWrapper(stream, 'ascii',
+                            errors='xmlcharrefreplace',
+                            line_buffering=True)
+
+
+def search_page(stream, form, where, logfile_pattern):
+    if "q" not in form:
+        print_search_form(stream)
+    else:
+        search_text = form["q"].value
+        if isinstance(search_text, bytes):
+            search_text = search_text.decode('UTF-8')
+        return print_search_results(search_text, stream=stream, where=where,
+                                    logfile_pattern=logfile_pattern)
+
+
+def get_path(environ):
+    path = environ.get('PATH_INFO', '/')
+    path = path.split('/')[-1]  # is this enough to protect from FS traversal?
+    return path if path else 'index.html'
+
+
+def wsgi(environ, start_response):
+    """WSGI application"""
+    logfile_path = environ.get('IRCLOG_LOCATION')
+    logfile_pattern = environ.get('IRCLOG_GLOB')
+    form = cgi.FieldStorage(fp=environ['wsgi.input'], environ=environ)
+    stream = io.TextIOWrapper(io.BytesIO(), 'ascii',
+                              errors='xmlcharrefreplace',
+                              line_buffering=True)
+
+    status = str("200 Ok")
+    content_type = str("text/html; charset=UTF-8")
+
+    path = get_path(environ)
+    if path == 'search':
+        fmt = search_page(stream, form, logfile_path, logfile_pattern)
+        result = [stream.buffer.getvalue()]
+    else:
+        if path.endswith('css'):
+            content_type = str("text/css")
+        try:
+            with open(os.path.join(logfile_path, path), "rb") as f:
+                result = [f.read()]
+        except IOError:
+            status = str("404 Not Found")
+            result = [b"Not found"]
+
+    headers = [(str("Content-Type"), content_type)]
+    start_response(status, headers)
+    return result
+
+def serve():
+    from wsgiref.simple_server import make_server
+    srv = make_server('localhost', 8080, wsgi)
+    print("Started at http://localhost:8080/")
+    srv.serve_forever()
```

Comment:
(That's if we decide to have an installed script for serving IRC logs, rather than moving this to buildout.cfg)